libexec/rtld-elf/rtld.c

```diff
@@ -4613,11 +4613,25 @@ do_dlsym(void *handle, const char *name, void *retaddr, const Ver_Entry *ve,
 	    sym = __DECONST(void*, make_function_pointer(def, defobj));
 	    dbg("dlsym(%s) is function: " PTR_FMT, name, sym);
 #ifdef CHERI_LIB_C18N
+	    /*
+	     * XXX Dapeng: Need to handle tail-calls causing the caller to be
+	     * mis-identified.
+	     */
+	    if (handle != NULL && handle != RTLD_NEXT &&
```

Comment:
I don't love duplicating all this, nor do I love reusing obj when it has another meaning for the handle case.

I'd suggest doing `retobj = obj_from_addr(retaddr)` before the big if above (with no error checking yet), and in the sentinel handle case do `obj = retobj`, checking for NULL as before. Then all you need here is to error if retobj is NULL.